drivers: serial: uart_altera_jtag: enhancement

gohshunjing opened the pull request:
implement uart poll in and interrupt driven api for uart_altera_jtag

Signed-off-by: Goh Shun Jing <shun.jing.goh@intel.com>

dcpleung commented on drivers/serial/uart_altera_jtag.c:
```diff
@@ -58,16 +133,21 @@ static void uart_altera_jtag_poll_out(const struct device *dev,
-	data_val &= ~UART_DATA_MASK;
-	data_val |= c;
-	sys_write32(data_val, config->base + UART_ALTERA_JTAG_DATA_OFFSET);
+	sys_write8(c, config->base + UART_ALTERA_JTAG_DATA_OFFSET);
```
Are you sure this still works on nios2, as you have removed the masking?

gohshunjing replied:
Hi @dcpleung thank you for the feedback.
I have verified this with niosv.
According to the [user guide](https://www.intel.com/content/www/us/en/docs/programmable/683130/22-3/data-register-48671.html) the data register bit 15 to 32 is read only.
So the masking is not necessary.
I have also change to sys_write8 (instead of 32), so that only 8 bits are written.
I will test with nios2 and get back to you.

gohshunjing replied:
Hi, 
i have tested with nios2 by using the following samples:

- hello_world
- synchronization
- philosophers

The printk work as expected.

The following overlay is needed, as jtag uart is not enabled by default.
in prj.conf:
`CONFIG_UART_ALTERA_JTAG=y`
in altera_max10.overlay:
```
/ {
     chosen {
         zephyr,console = &jtag_uart;
         zephyr,shell-uart = &jtag_uart;
     };
};

&jtag_uart {
    status = "okay";
};
```

gohshunjing commented:
i would like to mention, that these enhancement is not valid for nios2 if CONFIG_UART_ALTERA_JTAG_HAL is enabled.
There are no HAL function, that can be used directly for interrupt driven api.
Is this ok?

dcpleung commented:
Nios2 didn't work with HAL so it is not taken away functionality. So it is fine.

gohshunjing committed:
drivers: serial: uart_altera_jtag: enhancement

implement uart poll in and interrupt driven api.

Signed-off-by: Goh Shun Jing <shun.jing.goh@intel.com>

gohshunjing committed:
CODEOWNERS: Add code owener for uart_altera_jtag.c

Add code owener for uart_altera_jtag.c

Signed-off-by: Goh Shun Jing <shun.jing.goh@intel.com>

gohshunjing committed:
dts: nios2: intel: nios2-qemu: add jtag interrupt

assign interrupt number 0 for jtag_uart.
Number can be found in soc/nios2/nios2-qemu/include/system.h
JTAG_UART_0_IRQ 0

Signed-off-by: Goh Shun Jing <shun.jing.goh@intel.com>
